File: agents/work/skills/pi/roadmap-workflow/scripts/roadmap_parse.py
#!/usr/bin/env python3

import argparse
import csv
import json
import os
import re
import sys
from collections import Counter
from datetime import datetime, timezone

DATE_RE = re.compile(r"^\d{1,2}/\d{1,2}/\d{4}$")


def normalize_text(value: str) -> str:
    if not value:
        return ""
    return " ".join(value.split())


def parse_date_key(date_value: str):
    m, d, y = [int(x) for x in date_value.split("/")]
    return (y, m, d)


def detect_header_and_start(rows):
    for i, row in enumerate(rows):
        if len(row) > 6 and normalize_text(row[0]).lower() == "quarters":
            return row, i + 1

    for i, row in enumerate(rows):
        date_count = 0
        for cell in row:
            cell = normalize_text(cell)
            if DATE_RE.match(cell):
                date_count += 1
        if date_count >= 5:
            return row, i + 1

    return rows[0] if rows else [], 1


def parse_csv(csv_path: str):
    with open(csv_path, newline="", encoding="utf-8-sig") as f:
        rows = list(csv.reader(f))

    header, start_index = detect_header_and_start(rows)

    date_columns = []
    for idx, cell in enumerate(header):
        date_value = normalize_text(cell)
        if DATE_RE.match(date_value):
            date_columns.append((idx, date_value))

    date_columns.sort(key=lambda pair: parse_date_key(pair[1]))

    unique_dates = []
    seen_dates = set()
    for _, date_value in date_columns:
        if date_value in seen_dates:
            continue
        seen_dates.add(date_value)
        unique_dates.append(date_value)

    items = []
    for raw_idx, row in enumerate(rows[start_index:], start=start_index + 1):
        quarter = normalize_text(row[0]) if len(row) > 0 else ""
        title = normalize_text(row[1]) if len(row) > 1 else ""
        goal = normalize_text(row[2]) if len(row) > 2 else ""
        dependencies = normalize_text(row[3]) if len(row) > 3 else ""
        lead = normalize_text(row[4]) if len(row) > 4 else ""

        if not title:
            continue

        timeline = []
        for idx, date_value in date_columns:
            if idx >= len(row):
                continue
            cell = normalize_text(row[idx])
            if not cell:
                continue
            timeline.append({"date": date_value, "text": cell})

        quarter_lower = quarter.lower()
        if "goal" in quarter_lower:
            row_kind = "goal"
        elif quarter_lower == "ongoing":
            row_kind = "ongoing"
        elif quarter:
            row_kind = "structured"
        else:
            row_kind = "context"

        item = {
            "source_row": raw_idx,
            "quarter": quarter,
            "title": title,
            "goal": goal,
            "dependencies": dependencies,
            "lead": lead,
            "timeline": timeline,
            "timeline_count": len(timeline),
            "is_goal_row": "goal" in quarter_lower,
            "row_kind": row_kind,
        }
        items.append(item)

    quarter_counts = Counter(item["quarter"] for item in items if item["quarter"])

    return {
        "generated_at": datetime.now(timezone.utc).isoformat(),
        "csv_path": os.path.abspath(csv_path),
        "item_count": len(items),
        "date_columns": unique_dates,
        "quarter_counts": dict(quarter_counts),
        "items": items,
    }


def load_json(path):
    with open(path, "r", encoding="utf-8") as f:
        return json.load(f)


def score_item(item, keyword_lower: str):
    score = 0
    if keyword_lower in item.get("title", "").lower():
        score += 6
    if keyword_lower in item.get("goal", "").lower():
        score += 3
    if keyword_lower in item.get("dependencies", "").lower():
        score += 2
    if keyword_lower in item.get("lead", "").lower():
        score += 1
    if keyword_lower in item.get("quarter", "").lower():
        score += 1

    for entry in item.get("timeline", []):
        if keyword_lower in entry.get("text", "").lower():
            score += 1

    return score


def cmd_parse(args):
    data = parse_csv(args.csv)
    if args.out:
        with open(args.out, "w", encoding="utf-8") as f:
            json.dump(data, f, ensure_ascii=False, indent=2)
    print(json.dumps({
        "ok": True,
        "action": "parse",
        "csv_path": data["csv_path"],
        "item_count": data["item_count"],
        "date_columns": data["date_columns"],
        "quarter_counts": data["quarter_counts"],
        "out": os.path.abspath(args.out) if args.out else None,
    }, ensure_ascii=False))


def cmd_summary(args):
    data = load_json(args.json)
    items = data.get("items", [])

    if args.quarter:
        quarter_lower = args.quarter.lower()
        items = [it for it in items if quarter_lower in it.get("quarter", "").lower()]

    timeline_items = [
        it
        for it in items
        if it.get("timeline_count", 0) > 0 and (args.include_context or it.get("row_kind") != "context")
    ]
    top_timeline = sorted(timeline_items, key=lambda it: it.get("timeline_count", 0), reverse=True)[: args.limit]

    quarter_counts = Counter(it.get("quarter", "") for it in items if it.get("quarter"))

    print(json.dumps({
        "ok": True,
        "action": "summary",
        "generated_at": data.get("generated_at"),
        "csv_path": data.get("csv_path"),
        "item_count": len(items),
        "date_columns": data.get("date_columns", []),
        "quarter_counts": dict(quarter_counts),
        "top_timeline_items": [
            {
                "quarter": it.get("quarter"),
                "title": it.get("title"),
                "lead": it.get("lead"),
                "timeline_count": it.get("timeline_count", 0),
            }
            for it in top_timeline
        ],
    }, ensure_ascii=False))


def cmd_timeline(args):
    data = load_json(args.json)
    keyword = args.keyword.strip()
    keyword_lower = keyword.lower()

    scored = []
    for item in data.get("items", []):
        if not args.include_context and item.get("row_kind") == "context":
            continue
        score = score_item(item, keyword_lower)
        if score > 0:
            scored.append((score, item))

    scored.sort(key=lambda pair: (-pair[0], pair[1].get("title", "")))
    matches = [
        {
            "score": score,
            "quarter": item.get("quarter"),
            "title": item.get("title"),
            "goal": item.get("goal"),
            "lead": item.get("lead"),
            "timeline": item.get("timeline", []),
        }
        for score, item in scored[: args.limit]
    ]

    print(json.dumps({
        "ok": True,
        "action": "timeline",
        "keyword": keyword,
        "match_count": len(scored),
        "matches": matches,
    }, ensure_ascii=False))


def cmd_search(args):
    data = load_json(args.json)
    keyword = args.keyword.strip()
    keyword_lower = keyword.lower()

    scored = []
    for item in data.get("items", []):
        if not args.include_context and item.get("row_kind") == "context":
            continue
        score = score_item(item, keyword_lower)
        if score > 0:
            scored.append((score, item))

    scored.sort(key=lambda pair: (-pair[0], pair[1].get("title", "")))

    matches = []
    for score, item in scored[: args.limit]:
        preview = []
        for entry in item.get("timeline", []):
            if keyword_lower in entry.get("text", "").lower():
                preview.append(entry)
        if not preview:
            preview = item.get("timeline", [])[:2]

        matches.append({
            "score": score,
            "quarter": item.get("quarter"),
            "title": item.get("title"),
            "goal": item.get("goal"),
            "lead": item.get("lead"),
            "timeline_preview": preview,
        })

    print(json.dumps({
        "ok": True,
        "action": "search",
        "keyword": keyword,
        "match_count": len(scored),
        "matches": matches,
    }, ensure_ascii=False))


def cmd_sprint(args):
    data = load_json(args.json)
    target = normalize_text(args.date)
    if not DATE_RE.match(target):
        print(json.dumps({
            "ok": False,
            "action": "sprint",
            "error": "Invalid date format. Use m/d/yyyy (e.g. 2/17/2026).",
        }, ensure_ascii=False))
        return

    matches = []
    for item in data.get("items", []):
        if not args.include_context and item.get("row_kind") == "context":
            continue
        for entry in item.get("timeline", []):
            if entry.get("date") == target:
                matches.append({
                    "quarter": item.get("quarter"),
                    "title": item.get("title"),
                    "goal": item.get("goal"),
                    "lead": item.get("lead"),
                    "date": target,
                    "text": entry.get("text"),
                })

    matches = matches[: args.limit]

    print(json.dumps({
        "ok": True,
        "action": "sprint",
        "date": target,
        "match_count": len(matches),
        "matches": matches,
        "available_dates": data.get("date_columns", []),
    }, ensure_ascii=False))


def build_parser():
    parser = argparse.ArgumentParser(description="Parse and query roadmap CSV exports")
    sub = parser.add_subparsers(dest="command", required=True)

    parse_p = sub.add_parser("parse", help="Parse roadmap CSV and optionally write JSON")
    parse_p.add_argument("--csv", required=True, help="Path to roadmap CSV file")
    parse_p.add_argument("--out", required=False, help="Output JSON path")
    parse_p.set_defaults(func=cmd_parse)

    summary_p = sub.add_parser("summary", help="Summarize parsed roadmap JSON")
    summary_p.add_argument("--json", required=True, help="Parsed roadmap JSON path")
    summary_p.add_argument("--quarter", required=False, help="Quarter filter (substring)")
    summary_p.add_argument("--limit", type=int, default=10, help="Max top timeline items")
    summary_p.add_argument("--include-context", action="store_true", help="Include non-goal context rows")
    summary_p.set_defaults(func=cmd_summary)

    timeline_p = sub.add_parser("timeline", help="Get timeline for initiative keyword")
    timeline_p.add_argument("--json", required=True, help="Parsed roadmap JSON path")
    timeline_p.add_argument("--keyword", required=True, help="Keyword to match")
    timeline_p.add_argument("--limit", type=int, default=10, help="Max matches")
    timeline_p.add_argument("--include-context", action="store_true", help="Include non-goal context rows")
    timeline_p.set_defaults(func=cmd_timeline)

    search_p = sub.add_parser("search", help="Search roadmap rows by keyword")
    search_p.add_argument("--json", required=True, help="Parsed roadmap JSON path")
    search_p.add_argument("--keyword", required=True, help="Keyword to match")
    search_p.add_argument("--limit", type=int, default=20, help="Max matches")
    search_p.add_argument("--include-context", action="store_true", help="Include non-goal context rows")
    search_p.set_defaults(func=cmd_search)

    sprint_p = sub.add_parser("sprint", help="Get items planned for a sprint date")
    sprint_p.add_argument("--json", required=True, help="Parsed roadmap JSON path")
    sprint_p.add_argument("--date", required=True, help="Sprint date in m/d/yyyy")
    sprint_p.add_argument("--limit", type=int, default=50, help="Max matches")
    sprint_p.add_argument("--include-context", action="store_true", help="Include non-goal context rows")
    sprint_p.set_defaults(func=cmd_sprint)

    return parser


def main():
    parser = build_parser()
    args = parser.parse_args()
    args.func(args)


if __name__ == "__main__":
    try:
        main()
    except Exception as exc:
        print(json.dumps({"ok": False, "error": str(exc)}, ensure_ascii=False))
        sys.exit(1)

File: agents/work/skills/pi/roadmap-workflow/scripts/roadmap_parse.ts
#!/usr/bin/env bun

import { Effect, pipe } from "effect";
import { mkdir, readFile, writeFile } from "node:fs/promises";
import path from "node:path";

const DATE_RE = /^\d{1,2}\/\d{1,2}\/\d{4}$/;

type RowKind = "goal" | "ongoing" | "structured" | "context";

type TimelineEntry = {
  date: string;
  text: string;
};

type RoadmapItem = {
  source_row: number;
  quarter: string;
  title: string;
  goal: string;
  dependencies: string;
  lead: string;
  timeline: Array<TimelineEntry>;
  timeline_count: number;
  is_goal_row: boolean;
  row_kind: RowKind;
};

type ParsedRoadmap = {
  generated_at: string;
  csv_path: string;
  item_count: number;
  date_columns: Array<string>;
  quarter_counts: Record<string, number>;
  items: Array<RoadmapItem>;
};

type SummaryResult = {
  ok: true;
  action: "summary";
  generated_at: string;
  csv_path: string;
  item_count: number;
  date_columns: Array<string>;
  quarter_counts: Record<string, number>;
  top_timeline_items: Array<{
    quarter: string;
    title: string;
    lead: string;
    timeline_count: number;
  }>;
};

type TimelineResult = {
  ok: true;
  action: "timeline";
  keyword: string;
  match_count: number;
  matches: Array<{
    score: number;
    quarter: string;
    title: string;
    goal: string;
    lead: string;
    timeline: Array<TimelineEntry>;
  }>;
};

type SearchResult = {
  ok: true;
  action: "search";
  keyword: string;
  match_count: number;
  matches: Array<{
    score: number;
    quarter: string;
    title: string;
    goal: string;
    lead: string;
    timeline_preview: Array<TimelineEntry>;
  }>;
};

type SprintResult =
  | {
      ok: false;
      action: "sprint";
      error: string;
    }
  | {
      ok: true;
      action: "sprint";
      date: string;
      match_count: number;
      matches: Array<{
        quarter: string;
        title: string;
        goal: string;
        lead: string;
        date: string;
        text: string;
      }>;
      available_dates: Array<string>;
    };

class CliError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "CliError";
  }
}

const normalizeText = (value: string | undefined): string => {
  if (!value) return "";
  return value
    .split(/\s+/)
    .filter((part) => part.length > 0)
    .join(" ");
};

const parseDateKey = (dateValue: string): [number, number, number] => {
  const [month, day, year] = dateValue.split("/").map((value) => Number.parseInt(value, 10));
  return [year, month, day];
};

const compareDateStrings = (left: string, right: string): number => {
  const [leftY, leftM, leftD] = parseDateKey(left);
  const [rightY, rightM, rightD] = parseDateKey(right);

  if (leftY !== rightY) return leftY - rightY;
  if (leftM !== rightM) return leftM - rightM;
  return leftD - rightD;
};

const parseCsvRows = (rawContent: string): Array<Array<string>> => {
  const content = rawContent.startsWith("\uFEFF") ? rawContent.slice(1) : rawContent;
  if (content.length === 0) return [];

  const rows: Array<Array<string>> = [];
  let row: Array<string> = [];
  let field = "";
  let inQuotes = false;

  const pushField = () => {
    row.push(field);
    field = "";
  };

  const pushRow = () => {
    rows.push(row);
    row = [];
  };

  for (let index = 0; index < content.length; index++) {
    const char = content[index];

    if (inQuotes) {
      if (char === '"') {
        const next = content[index + 1];
        if (next === '"') {
          field += '"';
          index += 1;
        } else {
          inQuotes = false;
        }
      } else {
        field += char;
      }
      continue;
    }

    if (char === '"') {
      inQuotes = true;
      continue;
    }

    if (char === ",") {
      pushField();
      continue;
    }

    if (char === "\n") {
      pushField();
      pushRow();
      continue;
    }

    if (char === "\r") {
      pushField();
      pushRow();
      if (content[index + 1] === "\n") {
        index += 1;
      }
      continue;
    }

    field += char;
  }

  pushField();

  const endsWithNewline = content.endsWith("\n") || content.endsWith("\r");
  const isTrailingEmptyRow = row.length === 1 && row[0] === "";
  if (!(endsWithNewline && isTrailingEmptyRow)) {
    rows.push(row);
  }

  return rows;
};

const detectHeaderAndStart = (
  rows: Array<Array<string>>,
): { header: Array<string>; startIndex: number } => {
  for (let index = 0; index < rows.length; index++) {
    const row = rows[index] ?? [];
    if (row.length > 6 && normalizeText(row[0]).toLowerCase() === "quarters") {
      return { header: row, startIndex: index + 1 };
    }
  }

  for (let index = 0; index < rows.length; index++) {
    const row = rows[index] ?? [];
    let dateCount = 0;

    for (const cell of row) {
      if (DATE_RE.test(normalizeText(cell))) {
        dateCount += 1;
      }
    }

    if (dateCount >= 5) {
      return { header: row, startIndex: index + 1 };
    }
  }

  return { header: rows[0] ?? [], startIndex: 1 };
};

const countBy = (values: Array<string>): Record<string, number> => {
  const counts = new Map<string, number>();

  for (const value of values) {
    if (!value) continue;
    const current = counts.get(value) ?? 0;
    counts.set(value, current + 1);
  }

  return Object.fromEntries(counts.entries());
};

const parseCsvData = (csvPath: string, content: string): ParsedRoadmap => {
  const rows = parseCsvRows(content);
  const { header, startIndex } = detectHeaderAndStart(rows);

  const dateColumns: Array<{ index: number; date: string }> = [];
  for (let index = 0; index < header.length; index++) {
    const dateValue = normalizeText(header[index]);
    if (DATE_RE.test(dateValue)) {
      dateColumns.push({ index, date: dateValue });
    }
  }

  dateColumns.sort((left, right) => compareDateStrings(left.date, right.date));

  const uniqueDates: Array<string> = [];
  const seenDates = new Set<string>();
  for (const { date } of dateColumns) {
    if (seenDates.has(date)) continue;
    seenDates.add(date);
    uniqueDates.push(date);
  }

  const items: Array<RoadmapItem> = [];

  for (let rowIndex = startIndex; rowIndex < rows.length; rowIndex++) {
    const row = rows[rowIndex] ?? [];
    const sourceRow = rowIndex + 1;

    const quarter = normalizeText(row[0]);
    const title = normalizeText(row[1]);
    const goal = normalizeText(row[2]);
    const dependencies = normalizeText(row[3]);
    const lead = normalizeText(row[4]);

    if (!title) continue;

    const timeline: Array<TimelineEntry> = [];
    for (const { index, date } of dateColumns) {
      if (index >= row.length) continue;
      const text = normalizeText(row[index]);
      if (!text) continue;
      timeline.push({ date, text });
    }

    const quarterLower = quarter.toLowerCase();
    let rowKind: RowKind;
    if (quarterLower.includes("goal")) {
      rowKind = "goal";
    } else if (quarterLower === "ongoing") {
      rowKind = "ongoing";
    } else if (quarter) {
      rowKind = "structured";
    } else {
      rowKind = "context";
    }

    items.push({
      source_row: sourceRow,
      quarter,
      title,
      goal,
      dependencies,
      lead,
      timeline,
      timeline_count: timeline.length,
      is_goal_row: quarterLower.includes("goal"),
      row_kind: rowKind,
    });
  }

  return {
    generated_at: new Date().toISOString(),
    csv_path: path.resolve(csvPath),
    item_count: items.length,
    date_columns: uniqueDates,
    quarter_counts: countBy(items.map((item) => item.quarter)),
    items,
  };
};

const scoreItem = (item: RoadmapItem, keywordLower: string): number => {
  let score = 0;

  if (item.title.toLowerCase().includes(keywordLower)) score += 6;
  if (item.goal.toLowerCase().includes(keywordLower)) score += 3;
  if (item.dependencies.toLowerCase().includes(keywordLower)) score += 2;
  if (item.lead.toLowerCase().includes(keywordLower)) score += 1;
  if (item.quarter.toLowerCase().includes(keywordLower)) score += 1;

  for (const timelineEntry of item.timeline) {
    if (timelineEntry.text.toLowerCase().includes(keywordLower)) {
      score += 1;
    }
  }

  return score;
};

const readText = (filePath: string) =>
  Effect.tryPromise({
    try: () => readFile(filePath, "utf8"),
    catch: (error) =>
      new CliError(
        `Failed to read ${filePath}: ${error instanceof Error ? error.message : String(error)}`,
      ),
  });

const writeJsonFile = (filePath: string, value: unknown) =>
  Effect.tryPromise({
    try: async () => {
      const absolutePath = path.resolve(filePath);
      await mkdir(path.dirname(absolutePath), { recursive: true });
      await writeFile(absolutePath, `${JSON.stringify(value, null, 2)}\n`, "utf8");
    },
    catch: (error) =>
      new CliError(
        `Failed to write ${filePath}: ${error instanceof Error ? error.message : String(error)}`,
      ),
  });

const loadJson = <T>(filePath: string): Effect.Effect<T, CliError> =>
  pipe(
    readText(filePath),
    Effect.flatMap((raw) =>
      Effect.try({
        try: () => JSON.parse(raw) as T,
        catch: (error) =>
          new CliError(
            `Failed to parse JSON from ${filePath}: ${error instanceof Error ? error.message : String(error)}`,
          ),
      }),
    ),
  );

const parseInteger = (value: string, flagName: string): number => {
  const parsed = Number.parseInt(value, 10);
  if (Number.isNaN(parsed)) {
    throw new CliError(`Invalid ${flagName} value: ${value}`);
  }
  return parsed;
};

const ensureStringFlag = (value: string | boolean | undefined, flagName: string): string => {
  if (typeof value !== "string" || value.trim().length === 0) {
    throw new CliError(`Missing ${flagName}`);
  }
  return value;
};

const parseFlags = (tokens: Array<string>) => {
  const flags: Record<string, string | boolean> = {};
  const positional: Array<string> = [];

  for (let index = 0; index < tokens.length; index++) {
    const token = tokens[index];

    if (!token.startsWith("--")) {
      positional.push(token);
      continue;
    }

    const equalIndex = token.indexOf("=");
    if (equalIndex > -1) {
      const key = token.slice(2, equalIndex);
      const value = token.slice(equalIndex + 1);
      flags[key] = value;
      continue;
    }

    const key = token.slice(2);
    const next = tokens[index + 1];
    if (next && !next.startsWith("--")) {
      flags[key] = next;
      index += 1;
    } else {
      flags[key] = true;
    }
  }

  return { flags, positional };
};

const ensureAllowedFlags = (
  flags: Record<string, string | boolean>,
  allowed: ReadonlySet<string>,
): void => {
  for (const key of Object.keys(flags)) {
    if (!allowed.has(key)) {
      throw new CliError(`Unknown option --${key}`);
    }
  }
};

type Command =
  | { tag: "parse"; csv: string; out?: string }
  | { tag: "summary"; json: string; quarter?: string; limit: number; includeContext: boolean }
  | { tag: "timeline"; json: string; keyword: string; limit: number; includeContext: boolean }
  | { tag: "search"; json: string; keyword: string; limit: number; includeContext: boolean }
  | { tag: "sprint"; json: string; date: string; limit: number; includeContext: boolean };

const parseCommand = (argv: Array<string>): Command => {
  if (argv.length === 0) {
    throw new CliError("Missing command. Expected one of: parse, summary, timeline, search, sprint.");
  }

  const [commandName, ...rest] = argv;
  const { flags, positional } = parseFlags(rest);

  if (positional.length > 0) {
    throw new CliError(`Unexpected positional arguments: ${positional.join(" ")}`);
  }

  if (commandName === "parse") {
    ensureAllowedFlags(flags, new Set(["csv", "out"]));
    const csv = ensureStringFlag(flags.csv, "--csv");
    const out = typeof flags.out === "string" ? flags.out : undefined;
    return { tag: "parse", csv, out };
  }

  if (commandName === "summary") {
    ensureAllowedFlags(flags, new Set(["json", "quarter", "limit", "include-context"]));
    const json = ensureStringFlag(flags.json, "--json");
    const quarter = typeof flags.quarter === "string" ? flags.quarter : undefined;
    const limit =
      typeof flags.limit === "string"
        ? parseInteger(flags.limit, "--limit")
        : flags.limit === true
          ? (() => {
              throw new CliError("Missing --limit value");
            })()
          : 10;
    const includeContext = "include-context" in flags;
    return { tag: "summary", json, quarter, limit, includeContext };
  }

  if (commandName === "timeline") {
    ensureAllowedFlags(flags, new Set(["json", "keyword", "limit", "include-context"]));
    const json = ensureStringFlag(flags.json, "--json");
    const keyword = ensureStringFlag(flags.keyword, "--keyword").trim();
    const limit =
      typeof flags.limit === "string"
        ? parseInteger(flags.limit, "--limit")
        : flags.limit === true
          ? (() => {
              throw new CliError("Missing --limit value");
            })()
          : 10;
    const includeContext = "include-context" in flags;
    return { tag: "timeline", json, keyword, limit, includeContext };
  }

  if (commandName === "search") {
    ensureAllowedFlags(flags, new Set(["json", "keyword", "limit", "include-context"]));
    const json = ensureStringFlag(flags.json, "--json");
    const keyword = ensureStringFlag(flags.keyword, "--keyword").trim();
    const limit =
      typeof flags.limit === "string"
        ? parseInteger(flags.limit, "--limit")
        : flags.limit === true
          ? (() => {
              throw new CliError("Missing --limit value");
            })()
          : 20;
    const includeContext = "include-context" in flags;
    return { tag: "search", json, keyword, limit, includeContext };
  }

  if (commandName === "sprint") {
    ensureAllowedFlags(flags, new Set(["json", "date", "limit", "include-context"]));
    const json = ensureStringFlag(flags.json, "--json");
    const date = ensureStringFlag(flags.date, "--date").trim();
    const limit =
      typeof flags.limit === "string"
        ? parseInteger(flags.limit, "--limit")
        : flags.limit === true
          ? (() => {
              throw new CliError("Missing --limit value");
            })()
          : 50;
    const includeContext = "include-context" in flags;
    return { tag: "sprint", json, date, limit, includeContext };
  }

  throw new CliError(`Unknown command: ${commandName}`);
};

const runParse = (command: Extract<Command, { tag: "parse" }>) =>
  pipe(
    readText(command.csv),
    Effect.map((content) => parseCsvData(command.csv, content)),
    Effect.tap((parsed) => (command.out ? writeJsonFile(command.out, parsed) : Effect.void)),
    Effect.map((parsed) => ({
      ok: true,
      action: "parse",
      csv_path: parsed.csv_path,
      item_count: parsed.item_count,
      date_columns: parsed.date_columns,
      quarter_counts: parsed.quarter_counts,
      out: command.out ? path.resolve(command.out) : null,
    })),
  );

const runSummary = (command: Extract<Command, { tag: "summary" }>): Effect.Effect<SummaryResult, CliError> =>
  pipe(
    loadJson<ParsedRoadmap>(command.json),
    Effect.map((data) => {
      let items = data.items;

      if (command.quarter) {
        const quarterLower = command.quarter.toLowerCase();
        items = items.filter((item) => item.quarter.toLowerCase().includes(quarterLower));
      }

      const timelineItems = items.filter(
        (item) => item.timeline_count > 0 && (command.includeContext || item.row_kind !== "context"),
      );

      const topTimelineItems = [...timelineItems]
        .sort((left, right) => right.timeline_count - left.timeline_count)
        .slice(0, command.limit);

      return {
        ok: true,
        action: "summary",
        generated_at: data.generated_at,
        csv_path: data.csv_path,
        item_count: items.length,
        date_columns: data.date_columns,
        quarter_counts: countBy(items.map((item) => item.quarter)),
        top_timeline_items: topTimelineItems.map((item) => ({
          quarter: item.quarter,
          title: item.title,
          lead: item.lead,
          timeline_count: item.timeline_count,
        })),
      } satisfies SummaryResult;
    }),
  );

const runTimeline = (command: Extract<Command, { tag: "timeline" }>): Effect.Effect<TimelineResult, CliError> =>
  pipe(
    loadJson<ParsedRoadmap>(command.json),
    Effect.map((data) => {
      const keywordLower = command.keyword.toLowerCase();

      const scored = data.items
        .filter((item) => command.includeContext || item.row_kind !== "context")
        .map((item) => ({ score: scoreItem(item, keywordLower), item }))
        .filter(({ score }) => score > 0)
        .sort((left, right) => {
          if (left.score !== right.score) return right.score - left.score;
          return left.item.title.localeCompare(right.item.title);
        });

      return {
        ok: true,
        action: "timeline",
        keyword: command.keyword,
        match_count: scored.length,
        matches: scored.slice(0, command.limit).map(({ score, item }) => ({
          score,
          quarter: item.quarter,
          title: item.title,
          goal: item.goal,
          lead: item.lead,
          timeline: item.timeline,
        })),
      } satisfies TimelineResult;
    }),
  );

const runSearch = (command: Extract<Command, { tag: "search" }>): Effect.Effect<SearchResult, CliError> =>
  pipe(
    loadJson<ParsedRoadmap>(command.json),
    Effect.map((data) => {
      const keywordLower = command.keyword.toLowerCase();

      const scored = data.items
        .filter((item) => command.includeContext || item.row_kind !== "context")
        .map((item) => ({ score: scoreItem(item, keywordLower), item }))
        .filter(({ score }) => score > 0)
        .sort((left, right) => {
          if (left.score !== right.score) return right.score - left.score;
          return left.item.title.localeCompare(right.item.title);
        });

      return {
        ok: true,
        action: "search",
        keyword: command.keyword,
        match_count: scored.length,
        matches: scored.slice(0, command.limit).map(({ score, item }) => {
          const timelinePreview = item.timeline.filter((entry) =>
            entry.text.toLowerCase().includes(keywordLower),
          );

          return {
            score,
            quarter: item.quarter,
            title: item.title,
            goal: item.goal,
            lead: item.lead,
            timeline_preview: timelinePreview.length > 0 ? timelinePreview : item.timeline.slice(0, 2),
          };
        }),
      } satisfies SearchResult;
    }),
  );

const runSprint = (command: Extract<Command, { tag: "sprint" }>): Effect.Effect<SprintResult, CliError> =>
  pipe(
    loadJson<ParsedRoadmap>(command.json),
    Effect.map((data) => {
      const target = normalizeText(command.date);
      if (!DATE_RE.test(target)) {
        return {
          ok: false,
          action: "sprint",
          error: "Invalid date format. Use m/d/yyyy (e.g. 2/17/2026).",
        };
      }

      const matches = data.items
        .filter((item) => command.includeContext || item.row_kind !== "context")
        .flatMap((item) =>
          item.timeline
            .filter((entry) => entry.date === target)
            .map((entry) => ({
              quarter: item.quarter,
              title: item.title,
              goal: item.goal,
              lead: item.lead,
              date: target,
              text: entry.text,
            })),
        )
        .slice(0, command.limit);

      return {
        ok: true,
        action: "sprint",
        date: target,
        match_count: matches.length,
        matches,
        available_dates: data.date_columns,
      };
    }),
  );

const runCommand = (command: Command): Effect.Effect<unknown, CliError> => {
  switch (command.tag) {
    case "parse":
      return runParse(command);
    case "summary":
      return runSummary(command);
    case "timeline":
      return runTimeline(command);
    case "search":
      return runSearch(command);
    case "sprint":
      return runSprint(command);
    default: {
      const _exhaustive: never = command;
      return Effect.fail(new CliError(`Unhandled command: ${String(_exhaustive)}`));
    }
  }
};

const program = pipe(
  Effect.sync(() => parseCommand(process.argv.slice(2))),
  Effect.flatMap(runCommand),
  Effect.tap((payload) => Effect.sync(() => console.log(JSON.stringify(payload)))),
);

await Effect.runPromise(program).catch((error) => {
  const message = error instanceof Error ? error.message : String(error);
  console.log(JSON.stringify({ ok: false, error: message }));
  process.exit(1);
});
